src/Microsoft.AspNetCore.Razor.Language/Experiemental/Utf8Directive.cs

```diff
@@ -0,0 +1,143 @@
+            builder.AddTargetExtension(new TargetExtension());
+            return builder;
+        }
+
+        private class Utf8Node : ExtensionIntermediateNode
+        {
+            public Utf8Node(HtmlContentIntermediateNode node)
+            {
+                for (var i = 0; i < node.Children.Count; i++)
+                {
+                    Children.Add(node.Children[i]);
+                }
+            }
+
+            public override IntermediateNodeCollection Children { get; } = new IntermediateNodeCollection();
+
+            public override void Accept(IntermediateNodeVisitor visitor)
+            {
+                if (visitor == null)
+                {
+                    throw new ArgumentNullException(nameof(visitor));
+                }
+
+                AcceptExtensionNode<Utf8Node>(this, visitor);
+            }
+
+            public override void WriteNode(CodeTarget target, CodeRenderingContext context)
+            {
+                if (target == null)
+                {
+                    throw new ArgumentNullException(nameof(target));
+                }
+
+                if (context == null)
+                {
+                    throw new ArgumentNullException(nameof(context));
+                }
+
+                var extension = target.GetExtension<TargetExtension>();
+                if (extension == null)
+                {
+                    ReportMissingCodeTargetExtension<TargetExtension>(context);
+                    return;
+                }
+
+                extension.WriteUtf8(context, this);
+            }
+        }
+
+        private class Pass : IntermediateNodePassBase, IRazorOptimizationPass
+        {
+            // Runs before directive removal
+            public override int Order => DefaultFeatureOrder + 49;
+
+            protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+            {
+                if (documentNode.Options.DesignTime)
+                {
+                    return;
+                }
+
+                var directives = documentNode.FindDirectiveReferences(Directive);
+                if (directives.Count == 0)
+                {
+                    // Not utf8-ized
+                    return;
+                }
+
+                // OK, rewrite all of the static content
+                var visitor = new Visitor();
+                visitor.Visit(documentNode);
+            }
+        }
+
+        private class Visitor : IntermediateNodeWalker
+        {
+            public override void VisitHtml(HtmlContentIntermediateNode node)
+            {
+                var index = Parent.Children.IndexOf(node);
+                Parent.Children[index] = new Utf8Node(node);
+            }
+        }
+
+        private class TargetExtension : ICodeTargetExtension
+        {
+            private readonly StringBuilder _builder = new StringBuilder();
+            private int _count;
```

Comment:
Worth having a `Dictionary<string, int>` (or `ConcurrentDictionary<string, int>`) to reuse repeated items? 

e.g. in your brief sample ` new byte[]{13, 10}` occurs multiple times

Reply:
It would deffo help if we move to using fields + `ReadOnlyMemory<>`

Reply:
Like, if we really care we'd just want to remove newlines.